tokio/docs/runtime-stats.md

```diff
@@ -0,0 +1,137 @@
+
+## Guide-level explanation
+
+Runtime stats will be exposed via a struct that is attainable via the `tokio::runtime::Handle`. Calling `Handle::stats()` will return a reference counted struct that contains raw stat values. Through this, there will be a `tokio-metrics` crate that converts these raw stats into proper aggregated metrics that can be consumed by end user metrics collection systems like the `metrics` crate.
+
+```rust=
+// Handle that can be acquired via `Runtime` or via `Handle::current`.
+let handle = Handle::current();
+let stats = handle.stats();
+
+// executor level stats.
+let executor = stats.executor();
+
+// per-worker stats via the executor.
+for worker in executor.workers() {
+    println!("{}", worker.steal_count());
+}
+
+println!("{}", executor.global_queue_depth());
+
+// io driver stats.
+let io = stats.io();
+
+println!("{}", io.sources_count());
+
+let timer = stats.timer();
+
+println!("{}", timer.entry_count());
+```
+
+### Executor
+
+The handle will provide stats for each worker thread or in the case of the single threaded executor will provide a single worker. This provides a detailed view into what each worker is doing. Providing the ability for the `tokio-metrics` crate to expose the stats aggregated as a single metric or as a per-worker metric.
+
+The values will be updated in batch from the executor to avoid needing to stream the data on every action. This should amortize the cost by only needing to emit stats at a specific executor wall clock time. Where the executor wall clock time is determinetd by a single executor tick rather than actual system time. This allows the collectors to observe the time and the stats to determine how long certain executor cycles took. This removes the need to acquire the time during executor cycles.
+
+Each worker will expose these stats, updated in batches:
+
+- Amount of futures executed
+- Amount of work steal attempts
+- Amount of parks
+- Amount of times local queue overflowed into the global queue
+- Max local queue depth
+- Min local queue depth
+- Avg local queue depth
+- Queue depth at time of batch emission
+- Amount of executor ticks (loop iterations)
+- Number of `block_in_place` tasks entered
+
+The main goal of this implementation is to allow a user to run this metrics collection at all times in production with minimal overhead to their application. This would allow users to alarm on any regressions and track how the runtime is performing.
+
+Some of the stats include min/max (specifically the queue depth stats) this is because the depth of the queues changes throughout the stats batch window. The value could start low, spike up during the middle of the window then come back down. To understand this behavior the executor stats module will aggregate the depth values to reduce the need to stream the values.
```

Comment:
Is there a particular reason why we choose min/max/avg over, say, percentiles?

Reply:
I think it's a question of the runtime overhead caused by this. Percentiles will require more work will collecting data, and you might not be able to use atomic counters anymore.

An external aggregator that consumes the stats (e.g. once per second) could still perform aggregation and do percentiles based on the amount of occurrences inside the sampling period.

However an external aggregator won't be able to capture min/max values in case there are peaks inside that sampling period. E.g. if you want to have a metric which is around "maximum tasks polled inside an executor iteration" and "minimum tasks polled", you couldn't get that if you just have counters of
- eventloop iterations
- tasks polled

I guess for tasks where we find those values useful, it makes sense to add them.

Otherwise it's probably easiest to just add always incrementing counters and let the external application do the diffing and aggregation. You can provide some helpers that allow like:

```rust
let mut last_stats = stats.executor();
loop {
    std::thread::delay(sampling_time);
    let stats = stats.executor();
    let delta_stats = stats.diff(last_stats);
    my_favorite_metric_system.aggregate_and_emit(delta_stats); // or potentially also the raw stats
    last_stats = stats;
```

E.g. we had issues in the past where some metrics that only had been emitted once per minute didn't show BPS spikes that happened inside some seconds and caused excessive packet drops. 

@LucioFranco Might be worthwhile to document that kind of periodic sampling system in the "guide" section, since there had been a few questions on how to use the thing.